### PR TITLE
MC-1268 async config update fixes

### PR DESCRIPTION
**Pull request by erosb**
Adds `ConfigurationService#updateAsync(String)` method that just returns the UUID of the started config update process.

Also some renamings: 
 * AbstractIdentifiedEvent -> AbstractConfigUpdateEvent
 * AbstractConfigUpdateEvent#uuid -> AbstractConfigUpdateEvent#configUpdateProcessId

Fixes MC-1268

### Related PRs:

 * [Client protocol PR](https://github.com/hazelcast/hazelcast-client-protocol/pull/415)
 * [EE PR](https://github.com/hazelcast/hazelcast-enterprise/pull/4685)

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ConfigurationService.java [hunk 1]
```diff
@@ -86,13 +87,28 @@
      */
     ConfigUpdateResult update(Config newConfig);
 
+    /**
+     * Starts a configuration process asynchronously. Updates the configuration with the given configuration. Updating means
+     * dynamically changing all the differences dynamically changeable.
+     *
+     * @param configPatch string representation of the config patch, to find any new dynamically changeable sub configs
+     * @return update result which includes added and ignored configurations
```
Comment (promanenko) [inline hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ConfigurationService.java:95]: @return description doesn't match the signature

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractConfigUpdateEvent.java [hunk 2]
```diff
@@ -21,21 +21,22 @@
 import java.util.Objects;
 import java.util.UUID;
 
-abstract class AbstractIdentifiedEvent extends AbstractEventBase {
-    private final UUID uuid;
+abstract class AbstractConfigUpdateEvent
```
Comment (promanenko) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractConfigUpdateEvent.java:24]: I prefer to keep the old name. Not all events identified by UUID are config update events, for instance all WanConsistencyCheck related events or WanSync events.

Comment (erosb) [reply hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractConfigUpdateEvent.java:24]: This was discussed with Ramiz on slack. The problem with `AbstractIdentifiedEvent` is that the UUID doesn't identify the event itself, rather the whole config update process. So the events emitted by the same `ConfigUpdater` have the same UUID.

On the other hand `AbstractWanAntiEntropyEventBase` are indeed unrelated to config update. In 45c3567b56 I re-added the previous `AbstractWanEvent` which was deleted as part of the dynamic config work. @ramizdundar can you please review this change?

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractWanConfigurationEventBase.java [hunk 3]
```diff
@@ -20,11 +20,11 @@
...
 import java.util.UUID;
 
-public abstract class AbstractWanConfigurationEventBase extends AbstractIdentifiedEvent {
+public abstract class AbstractWanConfigurationEventBase extends AbstractConfigUpdateEvent {
     private final String wanConfigName;
 
-    protected AbstractWanConfigurationEventBase(UUID uuid, String wanConfigName) {
-        super(uuid);
+    protected AbstractWanConfigurationEventBase(UUID configUpdateProcessId, String wanConfigName) {
+        super(configUpdateProcessId);
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractWanConfigurationEventBase.java:27]:
```suggestion
    protected AbstractWanConfigurationEventBase(UUID uuid, String wanConfigName) {
        super(uuid);
```

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractWanConfigurationEventBase.java [hunk 4]
```diff
@@ -20,11 +20,11 @@
 
 import java.util.UUID;
 
-public abstract class AbstractWanConfigurationEventBase extends AbstractIdentifiedEvent {
+public abstract class AbstractWanConfigurationEventBase extends AbstractConfigUpdateEvent {
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractWanConfigurationEventBase.java:23]:
```suggestion
public abstract class AbstractWanConfigurationEventBase extends AbstractWanEvent {
```

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractWanEvent.java [hunk 5]
```diff
@@ -18,14 +18,14 @@
...
 
-abstract class AbstractIdentifiedEvent extends AbstractEventBase {
+abstract class AbstractWanEvent extends AbstractEventBase {
     private final UUID uuid;
 
-    protected AbstractIdentifiedEvent(UUID uuid) {
-        this.uuid = Objects.requireNonNull(uuid, "UUID must not be null");
+    protected AbstractWanEvent(UUID uuid) {
+        assert uuid != null : "UUID must not be null";
+        this.uuid = uuid;
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractWanEvent.java:28]: What is the reason for this change?

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractWanSyncFinishedEvent.java [hunk 6]
```diff
@@ -25,9 +25,9 @@
     private final long recordsSynced;
     private final long durationSecs;
 
-    AbstractWanSyncFinishedEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName, long durationSecs,
-                                 long recordsSynced, int partitionsSynced) {
-        super(uuid, wanReplicationName, wanPublisherId, mapName);
+    AbstractWanSyncFinishedEvent(UUID configUpdateProcessId, String wanReplicationName, String wanPublisherId, String mapName,
+                                 long durationSecs, long recordsSynced, int partitionsSynced) {
+        super(configUpdateProcessId, wanReplicationName, wanPublisherId, mapName);
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractWanSyncFinishedEvent.java:30]:
```suggestion
    AbstractWanSyncFinishedEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName, long durationSecs,
                                  long recordsSynced, int partitionsSynced) {
        super(uuid, wanReplicationName, wanPublisherId, mapName);
```

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/WanConsistencyCheckFinishedEvent.java [hunk 7]
```diff
@@ -27,9 +27,9 @@
     private final int checkedCount;
     private final int entriesToSync;
 
-    public WanConsistencyCheckFinishedEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName,
-                                            int diffCount, int checkedCount, int entriesToSync) {
-        super(uuid, wanReplicationName, wanPublisherId, mapName);
+    public WanConsistencyCheckFinishedEvent(UUID configUpdateProcessId, String wanReplicationName, String wanPublisherId,
+                                            String mapName, int diffCount, int checkedCount, int entriesToSync) {
+        super(configUpdateProcessId, wanReplicationName, wanPublisherId, mapName);
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/WanConsistencyCheckFinishedEvent.java:32]:
```suggestion
    public WanConsistencyCheckFinishedEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName,
                                             int diffCount, int checkedCount, int entriesToSync) {
        super(uuid, wanReplicationName, wanPublisherId, mapName);
```

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/WanConsistencyCheckIgnoredEvent.java [hunk 8]
```diff
@@ -30,9 +30,9 @@ public WanConsistencyCheckIgnoredEvent(String wanReplicationName, String wanPubl
         this(UuidUtil.newUnsecureUUID(), wanReplicationName, wanPublisherId, mapName, reason);
     }
 
-    public WanConsistencyCheckIgnoredEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName,
-                                           String reason) {
-        super(uuid, wanReplicationName, wanPublisherId, mapName);
+    public WanConsistencyCheckIgnoredEvent(UUID configUpdateProcessId, String wanReplicationName, String wanPublisherId,
+                                           String mapName, String reason) {
+        super(configUpdateProcessId, wanReplicationName, wanPublisherId, mapName);
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/WanConsistencyCheckIgnoredEvent.java:35]:
```suggestion
    public WanConsistencyCheckIgnoredEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName,
                                            String reason) {
        super(uuid, wanReplicationName, wanPublisherId, mapName);
```

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/WanConsistencyCheckStartedEvent.java [hunk 9]
```diff
@@ -22,9 +22,11 @@
...
 import static com.hazelcast.internal.management.events.EventMetadata.EventType.WAN_CONSISTENCY_CHECK_STARTED;
 
-public class WanConsistencyCheckStartedEvent extends AbstractWanAntiEntropyEventBase {
-    public WanConsistencyCheckStartedEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName) {
-        super(uuid, wanReplicationName, wanPublisherId, mapName);
+public class WanConsistencyCheckStartedEvent
+        extends AbstractWanAntiEntropyEventBase {
+    public WanConsistencyCheckStartedEvent(UUID configUpdateProcessId, String wanReplicationName, String wanPublisherId,
+                                           String mapName) {
+        super(configUpdateProcessId, wanReplicationName, wanPublisherId, mapName);
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/WanConsistencyCheckStartedEvent.java:29]:
```suggestion
public class WanConsistencyCheckStartedEvent extends AbstractWanAntiEntropyEventBase {
     public WanConsistencyCheckStartedEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName) {
         super(uuid, wanReplicationName, wanPublisherId, mapName);
```

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/WanSyncIgnoredEvent.java [hunk 10]
```diff
@@ -27,12 +27,12 @@
 public final class WanSyncIgnoredEvent extends AbstractWanAntiEntropyEventBase {
     private final String reason;
 
-    public WanSyncIgnoredEvent(UUID uuid,
+    public WanSyncIgnoredEvent(UUID configUpdateProcessId,
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/WanSyncIgnoredEvent.java:30]:
```suggestion
    public WanSyncIgnoredEvent(UUID uuid,
```

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/WanSyncIgnoredEvent.java [hunk 11]
```diff
@@ -27,12 +27,12 @@
...
                                String wanReplicationName,
                                String wanPublisherId,
                                String mapName,
                                String reason) {
-        super(uuid, wanReplicationName, wanPublisherId, mapName);
+        super(configUpdateProcessId, wanReplicationName, wanPublisherId, mapName);
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/WanSyncIgnoredEvent.java:35]:
```suggestion
        super(uuid, wanReplicationName, wanPublisherId, mapName);
```

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/WanSyncProgressUpdateEvent.java [hunk 12]
```diff
@@ -27,9 +27,9 @@
     private final int partitionsSynced;
     private final int recordsSynced;
 
-    public WanSyncProgressUpdateEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName,
-                                      int partitionsToSync, int partitionsSynced, int recordsSynced) {
-        super(uuid, wanReplicationName, wanPublisherId, mapName);
+    public WanSyncProgressUpdateEvent(UUID configUpdateProcessId, String wanReplicationName, String wanPublisherId,
+                                      String mapName, int partitionsToSync, int partitionsSynced, int recordsSynced) {
+        super(configUpdateProcessId, wanReplicationName, wanPublisherId, mapName);
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/WanSyncProgressUpdateEvent.java:32]:
```suggestion
    public WanSyncProgressUpdateEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName,
                                       int partitionsToSync, int partitionsSynced, int recordsSynced) {
        super(uuid, wanReplicationName, wanPublisherId, mapName);
```

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/WanSyncStartedEvent.java [hunk 13]
```diff
@@ -23,8 +23,8 @@
 import static com.hazelcast.internal.management.events.EventMetadata.EventType.WAN_SYNC_STARTED;
 
 public class WanSyncStartedEvent extends AbstractWanAntiEntropyEventBase {
-    public WanSyncStartedEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName) {
-        super(uuid, wanReplicationName, wanPublisherId, mapName);
+    public WanSyncStartedEvent(UUID configUpdateProcessId, String wanReplicationName, String wanPublisherId, String mapName) {
+        super(configUpdateProcessId, wanReplicationName, wanPublisherId, mapName);
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/WanSyncStartedEvent.java:27]:
```suggestion
    public WanSyncStartedEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName) {
        super(uuid, wanReplicationName, wanPublisherId, mapName);
```

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractConfigUpdateEvent.java (hunk 2 above)
Comment (ramizdundar) [reply hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractConfigUpdateEvent.java:24]: Here is my understanding:

This PR:
1. AbstractIdentifiedEvent -> AbstractConfigUpdateEvent
2. Create AbstractWanEvent
3. Wan events now extends AbstractWanEvent

What I imagined first:
1. Create AbstractConfigUpdateEvent
2. ConfigUpdate events now extends AbstractConfigUpdateEvent

How the current state differs:
1. Git history of AbstractConfigUpdateEvent is the history of super class of wan events
2. There are some leftovers (I tried to mark all of them)
3. Semantically AbstractIdentifiedEvent -> AbstractWanEvent (although not visible, on git AbstractWanEvent will be a new file)

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractWanAntiEntropyEventBase.java [hunk 14]
```diff
@@ -20,13 +20,14 @@
...
+public abstract class AbstractWanAntiEntropyEventBase extends AbstractWanEvent {
     protected final String wanReplicationName;
     protected final String wanPublisherId;
     protected final String mapName;
 
-    AbstractWanAntiEntropyEventBase(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName) {
-        super(uuid);
+    AbstractWanAntiEntropyEventBase(UUID configUpdateProcessId, String wanReplicationName, String wanPublisherId,
+                                    String mapName) {
+        super(configUpdateProcessId);
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractWanAntiEntropyEventBase.java:30]:
```suggestion
    AbstractWanAntiEntropyEventBase(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName) {
        super(uuid);
```

**Review by ramizdundar — COMMENTED** [review]
I left suggestions for few leftovers, please commit them with care (check style etc.), if you plan to use them.

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/WanConsistencyCheckFinishedEvent.java [hunk 15]
```diff
@@ -28,7 +28,7 @@
     private final int entriesToSync;
 
     public WanConsistencyCheckFinishedEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName,
-                                            int diffCount, int checkedCount, int entriesToSync) {
+                                             int diffCount, int checkedCount, int entriesToSync) {
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/WanConsistencyCheckFinishedEvent.java:31]:
```suggestion
                                            int diffCount, int checkedCount, int entriesToSync) {
```

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/WanConsistencyCheckIgnoredEvent.java [hunk 16]
```diff
@@ -31,7 +31,7 @@ public WanConsistencyCheckIgnoredEvent(String wanReplicationName, String wanPubl
     }
 
     public WanConsistencyCheckIgnoredEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName,
-                                           String reason) {
+                                            String reason) {
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/WanConsistencyCheckIgnoredEvent.java:34]:
```suggestion
                                           String reason) {
```

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/WanConsistencyCheckStartedEvent.java [hunk 17]
```diff
@@ -23,8 +23,8 @@
 import static com.hazelcast.internal.management.events.EventMetadata.EventType.WAN_CONSISTENCY_CHECK_STARTED;
 
 public class WanConsistencyCheckStartedEvent extends AbstractWanAntiEntropyEventBase {
-    public WanConsistencyCheckStartedEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName) {
-        super(uuid, wanReplicationName, wanPublisherId, mapName);
+     public WanConsistencyCheckStartedEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName) {
+         super(uuid, wanReplicationName, wanPublisherId, mapName);
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/WanConsistencyCheckStartedEvent.java:27]:
```suggestion
    public WanConsistencyCheckStartedEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName) {
        super(uuid, wanReplicationName, wanPublisherId, mapName);
```

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/WanSyncProgressUpdateEvent.java [hunk 18]
```diff
@@ -28,7 +28,7 @@
     private final int recordsSynced;
 
     public WanSyncProgressUpdateEvent(UUID uuid, String wanReplicationName, String wanPublisherId, String mapName,
-                                      int partitionsToSync, int partitionsSynced, int recordsSynced) {
+                                       int partitionsToSync, int partitionsSynced, int recordsSynced) {
```
Comment (ramizdundar) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/WanSyncProgressUpdateEvent.java:31]:
```suggestion
                                      int partitionsToSync, int partitionsSynced, int recordsSynced) {
```

**Review by ramizdundar — APPROVED** [review]
minor: few whitespace leftovers

**Review by promanenko — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ConfigurationService.java [hunk 19]
```diff
@@ -86,13 +87,29 @@
      */
     ConfigUpdateResult update(Config newConfig);
 
+    /**
+     * Starts a configuration process asynchronously. Updates the configuration with the given configuration. Updating means
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ConfigurationService.java:91]: Mİnor:

Maybe `Starts a configuration process` -> `Starts a configuration update process`?

Also, can you re-order those two so, instead of `update(Config), updateAsync(String), updateAsync(), update()`, we would have `update(Config), update(), updateAsync(String), updateAsync()`

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractWanEvent.java [hunk 20]
```diff
@@ -35,7 +36,7 @@ public UUID getUuid() {
     @Override
     public JsonObject toJson() {
         JsonObject json = new JsonObject();
-        json.add("uuid", uuid.toString());
+        json.add("uuid", uuid != null ? uuid.toString() : "null");
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractWanEvent.java:39]: Can this be null, as we have a check in the constructor?

**Review by mdumandag — APPROVED** [review]
LGTM, just have two minor comments

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ConfigurationService.java (hunk 19 above)
Comment (erosb) [reply hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ConfigurationService.java:91]: Sure, done.

File: hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractWanEvent.java (hunk 20 above)
Comment (erosb) [reply hazelcast/src/main/java/com/hazelcast/internal/management/events/AbstractWanEvent.java:39]: I removed the null check.

**Comment by erosb** [comment]
run-lab-run
